lib/federation.js

```diff
@@ -0,0 +1,396 @@
+  [Kind.INPUT_OBJECT_TYPE_DEFINITION]: Kind.INPUT_OBJECT_TYPE_EXTENSION
+}
+
+function removeExternalFields (typeDefinition) {
+  return {
+    ...typeDefinition,
+    fields: typeDefinition.fields.filter(
+      fieldDefinition =>
+        !fieldDefinition.directives.some(
+          directive => directive.name.value === 'external'
+        )
+    )
+  }
+}
+
+function getStubTypes (schemaDefinitions, isGateway) {
+  const definitionsMap = {}
+  const extensionsMap = {}
+  const extensions = []
+  const directiveDefinitions = []
+
+  for (const definition of schemaDefinitions) {
+    /* istanbul ignore next */
+    if (definition.kind === 'SchemaDefinition') {
+      continue
+    }
+
+    const typeName = definition.name.value
+    const isTypeExtensionByDirective = hasExtensionDirective(definition)
+    /* istanbul ignore else we are not interested in nodes that does not match if statements */
+    if (isTypeDefinitionNode(definition) && !isTypeExtensionByDirective) {
+      definitionsMap[typeName] = definition
+    } else if (
+      isTypeExtensionNode(definition) ||
+      (isTypeDefinitionNode(definition) && isTypeExtensionByDirective)
+    ) {
+      extensionsMap[typeName] = {
+        kind: isTypeExtensionByDirective
+          ? definition.kind
+          : extensionKindToDefinitionKind[definition.kind],
+        name: definition.name
+      }
+      if (isTypeExtensionByDirective) {
+        definition.kind = definitionKindToExtensionKind[definition.kind]
+      }
+      extensions.push(isGateway ? removeExternalFields(definition) : definition)
+    } else if (definition.kind === Kind.DIRECTIVE_DEFINITION) {
+      directiveDefinitions.push(definition)
+    }
+  }
+
+  return {
+    typeStubs: Object.keys(extensionsMap)
+      .filter(extensionTypeName => !definitionsMap[extensionTypeName])
+      .map(extensionTypeName => extensionsMap[extensionTypeName]),
+    extensions,
+    definitions: [...directiveDefinitions, ...Object.values(definitionsMap)]
+  }
+}
+
+function gatherDirectives (type) {
+  let directives = []
+  /* istanbul ignore else seems that extensionASTNodes is always provided in graphql16+, we are keeping this check for safety */
+  if (type.extensionASTNodes) {
+    for (const node of type.extensionASTNodes) {
+      /* istanbul ignore else we are not interested in nodes that does not have directives */
+      if (node.directives) {
+        directives = directives.concat(node.directives)
+      }
+    }
+  }
+
+  if (type.astNode && type.astNode.directives) {
+    directives = directives.concat(type.astNode.directives)
+  }
+
+  return directives
+}
+
+function typeIncludesDirective (type, directiveName) {
+  const directives = gatherDirectives(type)
+  return directives.some(directive => directive.name.value === directiveName)
+}
+
+function addTypeNameToResult (result, typename) {
+  /* istanbul ignore else when result is null or not an object we return original result */
```

Comment:
are we able to have tests covering the istanbul ignore else's in this file?

Reply:
I've removed some of them, others are not easy to remove and are there more to avoid future breaks.
I'll add a separated issue to address a proper investigation an find a case where they can be done.

Reply:
Perfect - sounds good to me!